src/sound.c

```diff
@@ -199,4 +205,155 @@ sound_free(void)
 }
 #endif
 
-#endif  // FEAT_SOUND && HAVE_CANBERRA
+#elif defined(MSWIN) // MSWIN
+
+static HWND g_hWndSound = NULL;
+
+    static LRESULT CALLBACK
+sound_wndproc(HWND hwnd, UINT message, WPARAM wParam, LPARAM lParam)
+{
+    soundcb_T	*p;
+
+    switch (message)
+    {
+	case MM_MCINOTIFY:
+	    for (p = first_callback; p != NULL; p = p->snd_next)
+		if (p->device_id == (MCIDEVICEID) lParam)
+		{
+		    typval_T	argv[3];
+		    typval_T	rettv;
+		    int		dummy;
+		    char	buf[16];
+
+		    vim_snprintf(buf, sizeof(buf), "close sound%06d", p->sound_id);
+		    mciSendString(buf, NULL, 0, 0);
+
+		    argv[0].v_type = VAR_NUMBER;
+		    argv[0].vval.v_number = p->sound_id;
+		    argv[1].v_type = VAR_NUMBER;
+		    argv[1].vval.v_number = wParam == MCI_NOTIFY_SUCCESSFUL ? 0
+					  : wParam == MCI_NOTIFY_ABORTED ? 1 : 2;
+		    argv[2].v_type = VAR_UNKNOWN;
+
+		    call_callback(&p->snd_callback, -1,
+					    &rettv, 2, argv, NULL, 0L, 0L, &dummy, TRUE, NULL);
+		    clear_tv(&rettv);
+
+		    delete_sound_callback(p);
+		    redraw_after_callback(TRUE);
+
+		}
+	    break;
+    }
+
+    return DefWindowProc(hwnd, message, wParam, lParam);
+}
+
+    static HWND
+sound_window()
+{
+    if (g_hWndSound == NULL)
+    {
+	LPCTSTR clazz = "VimSound";
+	WNDCLASS wndclass = { 0, sound_wndproc, 0, 0, g_hinst, NULL, 0, 0, NULL, clazz };
+	RegisterClass(&wndclass);
+	g_hWndSound = CreateWindow(clazz, NULL, 0, 0, 0, 0, 0,
+		HWND_MESSAGE, NULL, g_hinst, NULL);
+    }
+
+    return g_hWndSound;
+}
+
+    void
+f_sound_playevent(typval_T *argvars, typval_T *rettv)
+{
+    smsg("sound_playevent() not implemented");
+}
+
+    void
+f_sound_playfile(typval_T *argvars, typval_T *rettv)
+{
+    int		newid = sound_id + 1;
+    int		len;
+    char_u	*p, *esc;
+    WCHAR	*wp;
+    soundcb_T	*soundcb;
+    char	buf[32];
+    MCIERROR	err;
+
+    rettv->v_type = VAR_NUMBER;
+    rettv->vval.v_number = 0;
+
+    esc = vim_strsave_shellescape(
+		   tv_get_string(&argvars[0]), FALSE, FALSE);
```

Comment:
This line might be able to join with the previous line. It will fit in the 80 columns.